src/neo/IO/Caching/HashSetCache.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Neo.IO.Caching
+{
+    public class HashSetCache<T> : IReadOnlyCollection<T> where T : IEquatable<T>
+    {
+        /// <summary>
+        /// Sets where the Hashes are stored
+        /// </summary>      
+        private readonly LinkedList<HashSet<T>> sets = new LinkedList<HashSet<T>>();
+
+        /// <summary>
+        /// Maximum capacity of each bucket inside each HashSet of <see cref="sets"/>.
+        /// </summary>        
+        private readonly int bucketCapacity;
+
+        /// <summary>
+        /// Maximum number of buckets for the LinkedList, meaning its maximum cardinality.
+        /// </summary>
+        private readonly int maxBucketCount;
+
+        /// <summary>
+        /// Entry count
+        /// </summary>
+        public int Count { get; private set; }
+
+        public HashSetCache(int bucketCapacity, int maxBucketCount = 10)
+        {
+            if (bucketCapacity <= 0) throw new ArgumentOutOfRangeException($"{nameof(bucketCapacity)} should be greater than 0");
+            if (maxBucketCount <= 0) throw new ArgumentOutOfRangeException($"{nameof(maxBucketCount)} should be greater than 0");
+
+            this.Count = 0;
+            this.bucketCapacity = bucketCapacity;
+            this.maxBucketCount = maxBucketCount;
+            sets.AddFirst(new HashSet<T>());
+        }
+
+        public bool Add(T item)
+        {
+            if (Contains(item)) return false;
+            Count++;
+            if (sets.First.Value.Count < bucketCapacity) return sets.First.Value.Add(item);
+            var newSet = new HashSet<T>
+            {
+                item
+            };
+            sets.AddFirst(newSet);
+            if (sets.Count > maxBucketCount) sets.RemoveLast();
+            return true;
+        }
+
+        public bool Contains(T item)
+        {
+            foreach (var set in sets)
+            {
+                if (set.Contains(item)) return true;
+            }
+            return false;
+        }
+
+        public void ExceptWith(IEnumerable<T> items)
+        {
+            foreach (var item in items)
+            {
+                foreach (var set in sets)
```

Comment:
there is also another optimization, here @shargon and @eryeer.

We can invert the loop:

```cs
removed = a bool vector of removed items
foreach (var set in sets)
    foreach (var item in items)
        if(!removed{item})
           if (set.Remove(item))
               removed{item} = true
               if |removed| == |items| exit.
```

Reply:
If so, we need to create a Dictionary to store bool vector of removed items. If removing items is very large(in the performance test it can be as large as 50,000 tx), creating this dictionary will be very expensive and will cost additionary time.

Reply:
In fact, I though about a linear loop on items, iterating on it, then, we can just create a vector of flags, which perhaps less a quite low additional cost.

Reply:
On the other hand, I think this method has the same effect as the previous method. It must not be | removed | == | items | until all items have been traversed, right? Probably speaking, it can only be done after the complexity of O (itemCount * bucketCount), which is no different from the previous approach.

Reply:
@eryeer, this is aligned with @shargon suggestion.

We would create an interator over items starting from the most recent bucket.
As he mentioned, usually, the most recent hashes are the ones we are looking for, while the other older buckets may expire naturally.

In the worst case it would be `O(itemCount * bucketCount)`.
However, perhaps we could find all items already in the first set of items.

Reply:
I think that we can benchmark that in another PR, right?

Reply:
We are indeed traverse the items from the first set not the last. 😂

Reply:
You mean the default behavior of the `foreach`?

Reply:
Yes, because new item are always inserted in the first bucket.